routing/heap.go

```diff
@@ -40,6 +42,21 @@ type nodeWithDist struct {
 // algorithm to keep track of the "closest" node to our source node.
 type distanceHeap struct {
 	nodes []nodeWithDist
+
+	// pubkeyIndices maps public keys of nodes to their respective index in
+	// the heap. This is used as a way to avoid db lookups by using heap.Fix
+	// instead of having duplicate entries on the heap.
+	pubkeyIndices map[route.Vertex]int
```

Comment:
hmm... does this change have an actual performance impact? I would avoid adding this complexity if not.

Reply:
If `pubkeyIndices` isn't added, then if we have a duplicate entry to add to the heap (the second example I outlined way above), we can't call `heap.Fix` on it via `distanceHeap.PushOrFix`. If we get rid of that function, then the entry could potentially be the best entry in the heap and the `ForEachNodeChannel` loop is called on it.

Reply:
Does avoiding duplicates lead to a significant performance gain?

Reply:
I think the performance gain is noticeable, but these are milliseconds after all.
I'm colorblind so the non-red points are with these changes and the red points are on master without these changes.  Roughly 25% improvement in pathfinding.  The graph was exported from testnet with about 2500 nodes and about 8000 channels.  Then found a path to each node on the graph by modifying the json unmarshalling in `pathfind_test.go`.  Joost noticed a 15% improvement on QueryRoutes, but I think that includes some other sources of noise so I opted for this testing approach.

![pathfinding improvements](https://user-images.githubusercontent.com/15145615/59938289-93bdd000-9422-11e9-9ae2-0a1a903e351e.png)

Reply:
Here the red is master, the purple is with 446e873 and no later commits, and the non-red non-purple is with all the changes in the branch (removing duplicates). So there are incremental performance improvements when combining both approaches (converting the `LightningNode` to `Vertex` and removing duplicates)

![comparing with duplicates](https://user-images.githubusercontent.com/15145615/59939902-3bd59800-9427-11e9-9485-8e54726020ff.png)

Reply:
I tested out different source nodes and db sizes (minimal test db, medium 300mb, large 4gb) and concluded that both variables have the expected effect on time.  All the commits in the branch regularly outperform both master and commit 446e873.  The time gain varies depending on various conditions as well - heap memory usage etc but I found this to be the case across the board.  

Here in the first graph, the red is master and the non-red is with all the commits in the branch.  The master *mostly* stays within a band of 45ms - 90ms and seems to track both the 45ms and 90ms lines, std dev is high.  The non-red concentrates on a band between 30ms - 50ms with far lower std dev.  
<img width="1127" alt="2926-plot-1" src="https://user-images.githubusercontent.com/15145615/60145710-da9a2580-9794-11e9-9584-4dde96b102c4.png">

In the second graph, the purple represents 446e873 and clusters on the bottom around 30ms and at the top around 80ms but with std dev than the non-red non-purple points.  The purple points above the 50ms band are the points that get improved upon when deleting duplicates.
<img width="1122" alt="446e873_visited-2926" src="https://user-images.githubusercontent.com/15145615/60145723-e685e780-9794-11e9-82c8-a15b51f3e5af.png">

It is hard to say specifically what the % improvement is over master and over not taking duplicates into account since there are many variables that can affect this.  However, I do think that this performance improvement is worth the change.

Reply:
So this PR contains two speed ups:
1. Remove `OtherNode` db lookup
2. Don't process nodes from the heap that we already processed before. This saves an unnecessary lookup of channel edges from the db.

For 2, two options were tried:
2a. Add a visited map
2b. Use `heap.Fix`

I think 2a and 2b have more or less the same effect on performance, but it seems 2b is more efficient in terms of memory usage.

I also prefer 2b because it stays closer to the general descriptions given of the Dijkstra algorithm.

Reply:
@Crypt-iQ Thanks for the detailed performance stats! 😄